### ISSUE TITLE
Store GPS failure

### ISSUE THREAD

**Issue by SCdF** [medic/medic-webapp#2670]
It would be helpful to store the GPS failure reason in documents so we know why GPS wasn't acquired (at least vaguely).

Currently [we are storing failures](https://github.com/medic/medic-webapp/blob/master/static/js/services/enketo-prepopulation-data.js#L38-L39) (if the android wrapper comes back an error it will be valid JSON which we then set into Enketo), but it's being lost later, probably intentionally.

**Comment by SCdF** [comment]
@garethbowen / @alxndrsn any idea if this is intentional (not storing GPS failure errors in documents)?

**Comment by SCdF** [comment]
Another way of dealing with getting this information would be an implementation of #2031. A GPS failure would be a logable item.

**Comment by garethbowen** [comment]
It'd be nice to store it on the record somehow so it's clear why it's missing. I have no idea whether this was intentional or not.

**Comment by SCdF** [comment]
So I misunderstood the `bindJsonToXml` function: it takes your JSON and silently attempts to match it to whatever XML structure already exists. So anything that isn't a lat and a long (ie error reasons) are discarded.

**Comment by SCdF** [comment]
![image](https://cloud.githubusercontent.com/assets/583851/18079664/da428050-6e89-11e6-9a4e-6063ab608607.png)
Example of error data structure

**Comment by SCdF** [comment]
@garethbowen so I'm not sure what the most correct way to deal with this. A few different ideas:
 - Deal with this externally to the application: change forms so that the XML structure for location looks like this: `<location><lat /><long /><error /><message />`
 - Similar but different: change forms so that there is a separate XML structure elsewhere that we can bind to in case of error (though it's not clear where that would be apart from the above honestly)
 - Deal with this in the code: if there is a `<location>` element and we get an error, forcefully replace it with `<location><error>true</error><message>...</message></location>`
 - .. something else?

Which of these do you think is the most "correct" based on the rest of the app's approach to things?

**Comment by garethbowen** [comment]
I like the first option because it's explicit and doesn't require special code. I can't think of any better option...

**Comment by SCdF** [comment]
Ok, I made a minor change to generate an error if you're not on Android.

If forms are changed to accommodate those data structures errors will get stored correctly. Forms should be changed like this:
```diff
               <location>
                 <lat/>
                 <long/>
+                <error/>
+                <message/>
               </location>
```

@abbyad what would the next steps be here? I can just change the XML in projects for LG, but I'm probably not changing exactly the right thing and I don't want to mess up the process.

**Comment by sglangevin** [comment]
@SCdF @abbyad I think we would want to make the change in the `convert.sh` script so that this is included in every form, correct?

**Comment by SCdF** [comment]
_(looks right to me. Knew there was a reason I shouldn't just edit all of the xml files)_

**Comment by sglangevin** [comment]
@abbyad I've made the update to `convert.sh` and pushed to `medic-projects` after confirming that it creates the correct xml. Feel free to take a look.

**Comment by fredatthehub** [comment]
hi @SCdF , @abbyad, does the change suggested by @sglangevin mean that we can now start recording GPS failure reason?

**Comment by sglangevin** [comment]
@fredatthehub it's not clear to me whether any web app feature was built here (it looks like potentially not, but @SCdF would need to confirm). If not, then as soon as we upload forms that have been converted using the new conversion script, we'll be able to see the errors. If there was a web app feature built, then that feature would need to be released, probably in v2.9, before we started being able to see the error messages.

**Comment by SCdF** [comment]
Nope, no code changes :-) All that needs to be done is for forms to be generated with the changed convert.sh, or whatever process is involved in pushing changed forms.

**Comment by sglangevin** [comment]
Ok cool - then we can get on updating the forms so that we expose the error message in JSON. @fredatthehub can you work with @derickl to run the conversion script on all LG Kenya forms (even if there are no other changes)? That will get the right XML into the forms so that we can store any error message related to GPS. I will do the same for LG Uganda. @henokgetachew we should also do this for all BRAC forms. (cc @Enock1990)

**Comment by ngaruko** [comment]
Hi @sglangevin , this looks like a relatively old issue. Can safely assume that it now working as expected and remove the AT label? Thanks

**Comment by sglangevin** [comment]
Yes, this is currently working in that we are storing the error messages.

**Comment by Enock1990** [comment]
@SCdF I'm currently having conversations on this with LG. Is it possible to explain what the error(failure) messages mean. For example in relation to what the CHP is doing what does` provide 'passive' did not provide a location` mean? They are having 40% of there CHPs recording GPS so it would be paramount if we can explain why the others are not from the 3 failure messages.

**Comment by SCdF** [comment]
@Enock1990 I'm not completely sure what they mean (Android generates them, we don't). The closest thing I could find by googling was ironically http://medicmobile.cloud.answerhub.com/questions/145/what-do-the-various-geolocation-collection-errors.html , which just describes what passive etc mean, not why they are not successfully providing location data.

For the CHWs that are not recording GPS, is it 100% of that CHW's reports don't have GPS, or is it actually that 40% of reports have gps information against them and 60% do not?

**Comment by sglangevin** [comment]
From my testing, all I could conclusively say is that Airplane mode definitely blocks GPS collection. If airplane mode is on, you'll get `Provider 'network' did not provide a location.`

From the charts that @michaelkohn put together, some CHPs report GPS data 100% of the time, some less than 100% of the time and some 0% of the time, so it varies a lot. It does sound like they should make sure CHPs aren't using airplane mode (which some might be doing to save battery) and that they should review the location settings as well to make sure they are set to high accuracy mode.

**Comment by SCdF** [comment]
This might be a bit heavy handed, but we can detect airplane mode: https://stackoverflow.com/questions/4319212/how-can-one-detect-airplane-mode-on-android

We could update LG's android wrapper to do this.

Alternatively, we could do something more comprehensive (but more work) purely on the web side. Something like:
 - If we are configured to store GPS…
 - …and we don't get valid GPS data back when we store a report
 - Note the generated report, and when it was generated, in a queue somewhere, and display a banner to the user at the top of the screen
 - This banner would have a link to popup a fullscreen explanation on how to make sure GPS is enabled
 - They would have a "done" button or something to indicate they have enabled it again. We could even do a test loop to say "no we still can't get it". They would also have a "I can't get GPS now" button in case they're inside or whatever and it's just too bad
 - Once we have got legit GPS again go through the reports in the queue and add the current location. Could even note how long it was delayed by
 - *Or*, just automatically try to get GPS every minute for 20 minutes and then give up and clear the queue

**Comment by SCdF** [comment]
It would be good for LG to go have a chat with users who have 0% GPS reporting and look at their phones though, it might just be that simple: they need training to know how to turn 3G off but not GPS for example.

**Comment by alxndrsn** [comment]
Is it possible that these users just never _enabled_ GPS on their phones?  It's really hard to do on the Y4.

**Comment by sglangevin** [comment]
I believe that LG has a pretty standardized phone setup process, but it's worth confirming. I think they decentralized the process at some point so perhaps some groups of CHPs didn't have GPS enabled? I don't think there are any full branches with no GPS data for any CHP in the branch though. But we could try to group by training class to see if there are classes without any GPS as that might point to a phone setup issue. I also don't think that CHPs have access to adjust their phone settings but maybe some of them have been able to?

@Enock1990 should be able to help answer some of these questions.

**Comment by ngaruko** [comment]
I guess this has to do with phone settings. Is anyone able to confirm with LG phones if that's the case so we can close this @Enock1990 ? @sglangevin

**Comment by Enock1990** [comment]
@ngaruko what LG does is set up phones from the HQ. They also lock the settings with a password so that the CHPs don't temper with the settings.

**Comment by ngaruko** [comment]
Thanks @Enock1990 I guess originally this issue was about 'Store GPS failure'. If there are some other GPS detection/reporting issues, I think we will need another ticket to avoid confusion. I will move this to 'ready' unless there is a different opinion @sglangevin @Enock1990
